src/BatchProcessor.ts

```diff
@@ -1,19 +1,245 @@
 /* Copyright © 2024 Seneca Project Contributors, MIT License. */
 
+import Inks from 'inks'
+import { Gubu } from 'gubu'
+
+const ALL = '*'
+const Types: Record<string, number> = {
+  'Number': 1,
+  'Boolean': 1,
+  'String': 1,
+  'Array': 1,
+  'Object': 1
+}
 
 type BatchProcessorOptionsFull = {
   debug: boolean
+  send: any
+  where: any
 }
 
 export type BatchProcessorOptions = Partial<BatchProcessorOptionsFull>
 
 
 // FEATURE: subsets of keys by dot separators.
+class Match {
+  all: boolean
+  patrun: any
+  
+  constructor(patrun: any, all: any = null) {
+    this.patrun = patrun
+    this.all = all
+  }
+  
+  set_all(all: boolean) {
+    this.all = all
+  }
+  
+  find(pattern: any) {
+    return this.patrun.find(pattern)
+  }
+  
+  add_pattern(pattern: any, item: any) {
+    return this.patrun.add(pattern, item)
+  }
+  
+}
+
+class Utility {
+
+  static evaluateMessage(seneca: any, ctx: any, out: any, msg: any, body: any = null) {
+    const Jsonic = seneca.util.Jsonic
+    
+    msg = Object.assign(
+      typeof msg == 'string' ? Jsonic(msg) : { ...msg }, body)
+    
+    // console.log(i++, msg)
+
+    let new_msg: any = {}
+    for(let key in msg) {
+      let type = msg[key].split('~').pop()
+      let value = (Inks as any).evaluate(msg[key], { out, ctx }, { sep: '~' })
+  
+      if(null != value && Types[type]) {
+        let validate = Gubu(eval(type))
+        validate(value)
+      }
+  
+      if(null != value) {
+        new_msg[key] = value
+      } else {
+        new_msg[key] = msg[key]
+      }
+  
+    }
+    
+    return new_msg
+  }
+  
+  static async workflowRun(seneca: any, msg: any, config: any, options: any, results: any) {
```

Comment:
let's refactor this as a preprocessing step, so that workflow exposes a single function `run` without any conditional logic at runtime

the preprocessing step can then be unit tested by validating the data structure it generates